--- ./Makefile.am.org	2011-10-08 15:33:24.000000000 +0200
+++ ./Makefile.am	2011-12-20 12:31:21.988362106 +0100
@@ -65,7 +65,7 @@ endif
 
 ####################
 radsecproxy.conf.5: $(srcdir)/radsecproxy.conf.5.xml
-	docbook2x-man $<
+	docbook2X2man $<
 
 # Build HTML version of radsecproxy.conf.5.  NOTE: Only tested with
 # 'openjade' package installed on Ubuntu 9.10.
--- ./configure.ac.org	2012-04-27 14:53:59.000000000 +0200
+++ ./configure.ac	2012-04-30 12:36:22.135875837 +0200
@@ -116,10 +116,11 @@ AM_CONDITIONAL(ENABLE_EXPERIMENTAL_DYNDI
                test "$enable_experimental_dyndisc" = "yes")
 
 
-AC_ARG_VAR([DOCBOOK2X_MAN], [docbook2x-man])
-AC_CHECK_PROG([DOCBOOK2X_MAN], [docbook2x-man], [yes])
+AC_ARG_VAR([DOCBOOK2X_MAN], [docbook2X2man])
+AC_CHECK_PROG([DOCBOOK2X_MAN], [docbook2X2man], [yes])
 AM_CONDITIONAL(HAVE_DOCBOOK2X_MAN, test "$DOCBOOK2X_MAN" = "yes")
 
+
 AC_SUBST(TARGET_CFLAGS)
 AC_SUBST(TARGET_LDFLAGS)
 AX_CHECK_SSL
